instance/models/logging_mixin.py

```diff
@@ -0,0 +1,115 @@
+# You should have received a copy of the GNU Affero General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+#
+"""
+Instance app models - Logging - Mixins
+"""
+
+# Imports #####################################################################
+
+import logging
+
+from django.db import models
+
+from instance.models.logging_utils import level_to_integer
+
+
+# Constants ###################################################################
+
+# TODO: Don't propagate exceptions & debug data to end users
+PUBLISHED_LOG_LEVEL_SET = ('info', 'warn', 'error', 'exception')
+
+
+# Logging #####################################################################
+
+logger = logging.getLogger(__name__)
+
+
+# Models ######################################################################
+
+class LoggerMixin(models.Model):
+    """
+    Logging facilities - Logs stored on the model & shared with the client via websocket
+    """
+    class Meta:
+        abstract = True
+
+    def log(self, level, text, **kwargs):
+        """
+        Log an entry text at a specified level
+        """
+        if self.pk is not None:
+            self.logentry_set.create(level=level, text=text.rstrip(), **kwargs)
+        else:
+            level_integer = level_to_integer(level)
+            text = '{} [Log not attached to instance, not saved yet]'.format(text)
+            logger.log(level_integer, text)
+
+
+class LoggerInstanceMixin(LoggerMixin):
+    """
+    Logging facilities - Instances
+    """
+    class Meta:
+        abstract = True
+
+    @property
+    def log_text(self):
+        """
+        Combines the instance and server log outputs in chronological order
+        Currently only supports one non-terminated server at a time
+        Returned as a text string
+        """
+        current_server = self.active_server_set.get()
+        server_logentry_set = current_server.logentry_set.filter(level__in=PUBLISHED_LOG_LEVEL_SET)\
+                                                         .order_by('pk')\
+                                                         .iterator()
+        instance_logentry_set = self.logentry_set.filter(level__in=PUBLISHED_LOG_LEVEL_SET)\
+                                                 .order_by('pk')\
+                                                 .iterator()
+
+        def next_instance_logentry():
+            """ Get the next log entry from the instance logs iterator """
+            try:
+                return next(instance_logentry_set)
+            except StopIteration:
+                return None
+
+        def next_server_logentry():
+            """ Get the next log entry from the server logs iterator """
+            try:
+                return next(server_logentry_set)
+            except StopIteration:
+                return None
+
+        log_text = ''
+        instance_logentry = next_instance_logentry()
+        server_logentry = next_server_logentry()
+
+        while True:
```

Comment:
@antoviaque what happens here is a `merge` step in `mergesort`. SO, the more clear implementation would be:

    instance_logenrty = next_instance_logentry()
    server_logenrty = next_server_logentry()

    while instance_logentry and server_logentry:
        if server_logentry.created < instance_logentry.created:
             # append server entry and advance iterator
        else:
             # append instance entry and advence iterator

     while server_logentry:
           # append server entry and advance iterator
     while instance_logentry:
          # append instance entry and advance iterator

Basically the first while consumes everything until one of the iterators is empty, than the rest of the code just appends anything left in the remaining iterator to the end. As you can see, much less ifs :)